compsequential.go

```diff
@@ -0,0 +1,240 @@
+		})
+}
+
+// If some router fails and the IgnoreError flag is true, we continue to the next router.
+// Context timeout error will be also ignored if the flag is set.
+func (r *composableSequential) GetValue(ctx context.Context, key string, opts ...routing.Option) ([]byte, error) {
+	return getValueOrErrorSequential(ctx, r.routers,
+		func(ctx context.Context, r routing.Routing) ([]byte, bool, error) {
+			val, err := r.GetValue(ctx, key, opts...)
+			return val, len(val) == 0, err
+		},
+	)
+}
+
+// If some router fails and the IgnoreError flag is true, we continue to the next router.
+// Context timeout error will be also ignored if the flag is set.
+func (r *composableSequential) SearchValue(ctx context.Context, key string, opts ...routing.Option) (<-chan []byte, error) {
+	return getChannelOrErrorSequential(ctx, r.routers,
+		func(ctx context.Context, r routing.Routing) (<-chan []byte, error) {
+			return r.SearchValue(ctx, key, opts...)
+		},
+		func() bool { return false },
+	)
+
+}
+
+// If some router fails and the IgnoreError flag is true, we continue to the next router.
+// Context timeout error will be also ignored if the flag is set.
+func (r *composableSequential) Bootstrap(ctx context.Context) error {
+	return executeSequential(ctx, r.routers,
+		func(ctx context.Context, r routing.Routing) error {
+			return r.Bootstrap(ctx)
+		},
+	)
+}
+
+func getValueOrErrorSequential[T any](
+	ctx context.Context,
+	routers []*SequentialRouter,
+	f func(context.Context, routing.Routing) (T, bool, error),
+) (value T, err error) {
+	for _, router := range routers {
+		ctx, cancel := context.WithTimeout(ctx, router.Timeout)
+		defer cancel()
+		value, empty, err := f(ctx, router.Router)
+		if err != nil &&
+			!errors.Is(err, routing.ErrNotFound) &&
+			!router.IgnoreError {
+			return value, err
+		}
+
+		if empty {
+			continue
+		}
+
+		return value, nil
+	}
+
+	return value, routing.ErrNotFound
+}
+
+func executeSequential(
+	ctx context.Context,
+	routers []*SequentialRouter,
+	f func(context.Context, routing.Routing,
+	) error) error {
+	for _, router := range routers {
+		ctx, cancel := context.WithTimeout(ctx, router.Timeout)
+		if err := f(ctx, router.Router); err != nil &&
+			!errors.Is(err, routing.ErrNotFound) &&
+			!router.IgnoreError {
+			cancel()
+			return err
+		}
+		cancel()
+	}
+
+	return nil
+}
+
+func getChannelOrErrorSequential[T any](
+	ctx context.Context,
+	routers []*SequentialRouter,
+	f func(context.Context, routing.Routing) (<-chan T, error),
+	shouldStop func() bool,
+) (chan T, error) {
+	chanOut := make(chan T)
+	var chans []<-chan T
+	var cancels []context.CancelFunc
+
+	for _, router := range routers {
+		ctx, cancel := context.WithTimeout(ctx, router.Timeout)
+		rch, err := f(ctx, router.Router)
+		if err != nil &&
+			!errors.Is(err, routing.ErrNotFound) &&
+			!router.IgnoreError {
+			cancel()
+			return nil, err
+		}
+
+		cancels = append(cancels, cancel)
+		chans = append(chans, rch)
+	}
```

Comment:
yeah, you are right, this is strange behavior. Changing it.

Reply:
Actually, there is no other way to do it. We need to check all the errors before starting to iterate channels sequentially. Example: we cannot return an error for the second router if the first one started filling the channel. At the channel level, we are reading sequentially. Maybe I didn't get something.

Reply:
FindProvidersAsync doesn't return errors anyway, if there's an error and IgnoreErrors=true then we continue to the next router, otherwise we just close the chan and return, right?

Reply:
Added the change here: [`b95a96a` (#58)](https://github.com/libp2p/go-libp2p-routing-helpers/pull/58/commits/b95a96ae8fb1f27bd4522d5925a99fbb6035e11a) this will swallow the error, but it will continue until a router fails getting the channel and ignoreErrors is false.